### PR TITLE
Add fields `ExtraArgsArray` and `WindowsExtraArgsArray`

### PR DESCRIPTION
**Pull request by HarrisonWAffel**
Issue: https://github.com/rancher/rancher/issues/25500
 
# Problem
When setting up [service account token volume projection](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-token-volume-projection) within the cluster.yaml, users were not able to specify a flag multiple times with different values within the `ExtraArgs` section of the `kubeapi` service. This prevents users from specifying multiple `service-account-key-file`’s, multiple `service-account-issuer`’s, or multiple `api-audiences`. This issue may also exist for the other services, and is not specific to service account projection (although that is the requested usecase)

# Solution
In an effort to preserve backwards compatibility, two new fields have been added to the `BaseService`: `ExtraArgsArray`/`extra_args_array` & `WindowsExtraArgsArray`/`win_extra_args_array`. These fields have a type of `map[string][]string`, as opposed to the type of `map[string]string` which is used for the existing `extraArgs` field. These new fields allow users to specify arguments multiple times with different values. These new fields can be used within any of the services within the cluster.yaml. _If anyone has any better names for these two fields please let me know, I'm happy to change them._ 

# Testing
I was able to utilize the new field within the `kubeapi` service when creating or modifying a cluster. I was able to observe updates / failures when specifying valid or invalid values for the arguments. Here are the test cases I have performed:

Test case 1) 
 + Create a new DO cluster (1 node, all roles)
 + Once available, SSH into the node
   + run the following commands 
     + `cp /etc/kubernetes/ssl/kube-apiserver-key.pem /etc/kubernetes/ssl/kube-apiserver-key2.pem`
     + `cp /etc/kubernetes/ssl/kube-apiserver-key.pem /etc/kubernetes/ssl/kube-apiserver-key3.pem`
 + modify the cluster.yaml and replace the `kubeapi` section with the following 
```
kube-api:
      always_pull_images: false
      extra_args_array:
        service-account-key-file:
          - /etc/kubernetes/ssl/kube-apiserver-key.pem
          - /etc/kubernetes/ssl/kube-apiserver-key2.pem
          - /etc/kubernetes/ssl/kube-apiserver-key3.pem
      pod_security_policy: false
      secrets_encryption_config:
        enabled: false
      service_node_port_range: 30000-32767
```
 + apply changes and observe cluster go into `active` state

Test case 2) 
 + create new downstream cluster in DO (1 node, all roles)
 + modify the cluster yaml BEFORE creating the cluster
 + Replace the `kube-api` field with the following value 
```kube-api:
      always_pull_images: false
      extra_args_array:
        service-account-key-file:
          - /etc/kubernetes/ssl/kube-apiserver-key.pem
          - /etc/kubernetes/ssl/blah.pem
      pod_security_policy: false
      secrets_encryption_config:
        enabled: false
      service_node_port_range: 30000-32767
```
+ Save the cluster 
+ Observe cluster errors out with the following error (expected) ```
[controlPlane] Failed to upgrade Control Plane: [[Failed to verify healthcheck: Failed to check https://localhost:6443/healthz for service [kube-apiserver] on host [104.131.53.240]: Get "https://localhost:6443/healthz": dial tcp [::1]:6443: connect: connection refused, log: Error: open /etc/kubernetes/ssl/blah.pem: no such file or directory]]```

If there are other test cases I should attempt before merging this PR let me know!

### CODE REVIEW

File: cluster/plan.go [hunk 1]
```diff
@@ -392,6 +393,12 @@ func (c *Cluster) BuildKubeControllerProcess(host *hosts.Host, serviceOptions v3
...
 
+	for arg, value := range c.Services.KubeController.ExtraArgsArray {
+		CommandArgs[arg] = value
+	}
+
+	Command = appendArgs(Command, CommandArgs)
+
 	for arg, value := range CommandArgs {
 		cmd := fmt.Sprintf("--%s=%s", arg, value)
 		Command = append(Command, cmd)
```
Comment (jiaqiluo) [inline cluster/plan.go]: It looks like the above call to appenArgs has done the work, so we can drop this section.

File: cluster/plan.go [hunk 2]
```diff
@@ -596,6 +603,12 @@ func (c *Cluster) BuildKubeletProcess(host *hosts.Host, serviceOptions v3.Kubern
 		CommandArgs[arg] = value
 	}
 
+	for arg, value := range host.GetExtraArgsArray(kubelet.BaseService) {
+		CommandArgs[arg] = value
+	}
+
+	Command = appendArgs(Command, CommandArgs)
```
Comment (jiaqiluo) [inline cluster/plan.go:612]: we should drop this line because the same function will be called at line 624 
```
Command: appendArgs(Command, CommandArgs),
```

File: cluster/plan.go [hunk 3]
```diff
@@ -1044,8 +1066,16 @@ func (c *Cluster) BuildEtcdProcess(host *hosts.Host, etcdHosts []*hosts.Host, se
...
+		case []string:
+			for _, v := range value.([]string) {
+				cmd := fmt.Sprintf("--%s=%s", arg, v)
+				args = append(args, cmd)
+			}
+		case string:
+			cmd := fmt.Sprintf("--%s=%s", arg, value)
+			args = append(args, cmd)
+		}
 	}
```
Comment (jiaqiluo) [inline cluster/plan.go]: can we simplify this section like the below?
```
args := appendArgs(args, CommaArgs)
```

File: hosts/hosts.go [hunk 4]
```diff
@@ -420,6 +420,15 @@ func (h *Host) GetExtraArgs(service v3.BaseService) map[string]string {
 	}
 }
 
+func (h *Host) GetExtraArgsArray(service v3.BaseService) map[string][]string {
+	switch {
+	case h.OS() == "windows" && len(service.WindowsExtraArgsArray) > 0:
```
Comment (jiaqiluo) [inline hosts/hosts.go]: If the `ExtraArgsArray` is supposed to be used when the OS is NOT windows, in the case where we have `ExtraArgsArray` but not `WindowsExtraArgsArray`, will it return the `ExtraArgsArray`? 

```suggestion
	case h.OS() == "windows":
```

**Review by jiaqiluo — CHANGES_REQUESTED** [review]

File: hosts/hosts.go (hunk 4 above)
Comment (HarrisonWAffel) [reply hosts/hosts.go]: Yes it will, this functions in the same fashion as the function above it `GetExtraArgs`

File: cluster/plan.go [hunk 5]
```diff
@@ -953,7 +965,7 @@ func (c *Cluster) BuildEtcdProcess(host *hosts.Host, etcdHosts []*hosts.Host, se
 		listenAddress = "0.0.0.0"
 	}
 
-	CommandArgs := map[string]string{
+	CommandArgs := map[string]interface{}{
```
Comment (kinarashah) [inline cluster/plan.go]: Is there a particular reason why we want to convert string to interface here? I'd prefer keeping commandArgs and commandArgsArray separate to make it explicit vs having to use switch case on types. Also this'd help ensure we only need to test this new field. 

I do like how you updated it to use one function instead of repeating the same logic for all plans 👍🏼 (maybe we could have 2 funcs instead of one)

File: hosts/hosts.go (hunk 4 above)
Comment (jiaqiluo) [reply hosts/hosts.go]: I am ok with it if that is how we handle the same case at all places.

**Review by jiaqiluo — APPROVED** [review]
LGTM

File: cluster/plan.go [hunk 6]
```diff
@@ -188,6 +188,8 @@ func (c *Cluster) BuildKubeAPIProcess(host *hosts.Host, serviceOptions v3.Kubern
 		"tls-cert-file":                pki.GetCertPath(pki.KubeAPICertName),
 		"tls-private-key-file":         pki.GetKeyPath(pki.KubeAPICertName),
 	}
+	CommandArrayArgs := map[string][]string{}
```
Comment (thedadams) [inline cluster/plan.go]:
```suggestion
	CommandArrayArgs := make(map[string][]string, len(c.Services.KubeAPI.ExtraArgsArray))
```

File: cluster/plan.go [hunk 7]
```diff
@@ -339,6 +343,7 @@ func (c *Cluster) BuildKubeControllerProcess(host *hosts.Host, serviceOptions v3
 		"service-account-private-key-file": pki.GetKeyPath(pki.ServiceAccountTokenKeyName),
 		"service-cluster-ip-range":         c.Services.KubeController.ServiceClusterIPRange,
 	}
+	CommandArrayArgs := map[string][]string{}
```
Comment (thedadams) [inline cluster/plan.go]:
```suggestion
	CommandArrayArgs := make(map[string][]string, len(c.Services.KubeController.ExtraArgsArray))
```

File: cluster/plan.go [hunk 8]
```diff
@@ -451,6 +459,7 @@ func (c *Cluster) BuildKubeletProcess(host *hosts.Host, serviceOptions v3.Kubern
 		"pod-infra-container-image": kubelet.InfraContainerImage,
 		"root-dir":                  path.Join(host.PrefixPath, "/var/lib/kubelet"),
 	}
+	CommandArrayArgs := map[string][]string{}
```
Comment (thedadams) [inline cluster/plan.go]: Can we do something similar here as with my previous suggestions to save on unnecessary allocations?

File: cluster/plan.go [hunk 9]
```diff
@@ -633,6 +649,7 @@ func (c *Cluster) BuildKubeProxyProcess(host *hosts.Host, serviceOptions v3.Kube
 		"hostname-override": host.HostnameOverride,
 		"kubeconfig":        pki.GetConfigPath(pki.KubeProxyCertName),
 	}
+	CommandArrayArgs := map[string][]string{}
```
Comment (thedadams) [inline cluster/plan.go]: Can we do something similar here as with my previous suggestions to save on unnecessary allocations?

File: cluster/plan.go [hunk 10]
```diff
@@ -786,7 +810,7 @@ func (c *Cluster) BuildSchedulerProcess(host *hosts.Host, serviceOptions v3.Kube
 	CommandArgs := map[string]string{
 		"kubeconfig": pki.GetConfigPath(pki.KubeSchedulerCertName),
 	}
-
+	CommandArrayArgs := map[string][]string{}
```
Comment (thedadams) [inline cluster/plan.go]:
```suggestion
	CommandArrayArgs := make(map[string][]string, len(c.Services.Scheduler.ExtraArgsArray))
```

File: cluster/plan.go [hunk 11]
```diff
@@ -969,6 +995,7 @@ func (c *Cluster) BuildEtcdProcess(host *hosts.Host, etcdHosts []*hosts.Host, se
 		"peer-cert-file":              pki.GetCertPath(nodeName),
 		"peer-key-file":               pki.GetKeyPath(nodeName),
 	}
+	CommandArrayArgs := map[string][]string{}
```
Comment (thedadams) [inline cluster/plan.go]:
```suggestion
	CommandArrayArgs := make(map[string][]string, len(c.Services.Etcd.ExtraArgsArray))
```

File: hosts/hosts.go [hunk 12]
```diff
@@ -420,6 +420,15 @@ func (h *Host) GetExtraArgs(service v3.BaseService) map[string]string {
...
+		return service.WindowsExtraArgsArray
+	default:
+		return service.ExtraArgsArray
+	}
+}
```
Comment (thedadams) [inline hosts/hosts.go]: nit: switch is not necessary here.
```suggestion
	if h.OS() == "windows" && len(service.WindowsExtraArgsArray) > 0 {
		return service.WindowsExtraArgsArray
	}

	return service.ExtraArgsArray
}
```

File: types/zz_generated_deepcopy.go [hunk 13]
```diff
@@ -249,6 +249,21 @@ func (in *BaseService) DeepCopyInto(out *BaseService) {
 			(*out)[key] = val
 		}
 	}
+	if in.ExtraArgsArray != nil {
```
Comment (thedadams) [inline types/zz_generated_deepcopy.go:252]: `go generate` results should be in a separate commit.

**Review by thedadams — CHANGES_REQUESTED** [review]

File: types/zz_generated_deepcopy.go (hunk 13 above)
Comment (HarrisonWAffel) [reply types/zz_generated_deepcopy.go:252]: fixed

**Review by thedadams — APPROVED** [review]
Looks good. Please don't squash these commits on merge.

**Review by jiaqiluo — APPROVED** [review]
LGTM, 
before merging, could you run the tests again to make sure it still works as expected?

**Review by jakefhyde — APPROVED** [review]
